# GCC118 - Programação Matemática
## Universidade Federal de Lavras
### Problema 1
#### Aluno: Eduardo Cesar Cauduro Coelho
#### Matrícula: 202310175
#### Aluno: Felipe Geraldo de Oliveira
#### Matrícula: 202310174

## Instalação da biblioteca Gurobipy





In [1]:
!pip install gurobipy

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 14.4/14.4 MB 18.6 MB/s eta 0:00:00


# 1$)$ Implemente um modelo de programação linear, usando o gurobipy, e encontre a solução ótima do problema:

## Declaração dos parâmetros do relógio de parede e do relógio de pedestal, além do número máximo de horas que cada pessoa trabalha



In [2]:
relogio_pedestal = {
    "lucro": 300,
    "tempo_gasto_montagem": 6,
    "tempo_gasto_producao": 8,
    "tempo_gasto_remessa": 3
}

relogio_parede = {
    "lucro": 200,
    "tempo_gasto_montagem": 4,
    "tempo_gasto_producao": 4,
    "tempo_gasto_remessa": 3
}

horas_maximas_trabalhadas = {
    "Joao": 40,
    "Ana": 40,
    "Lidia": 20
}

## Criação do modelo:

In [3]:
from gurobipy import Model, GRB

modelo = Model("relogios")

Restricted license - for non-production use only - expires 2026-11-23


## Declaração das variáveis de decisão que indicam o número de relógios de pedestal vendidos e o número de relógios de parede vendidos

In [4]:
relogio_pedestal_var = modelo.addVar(name="relogio_pedestal", vtype=GRB.CONTINUOUS, lb=0)
relogio_parede_var = modelo.addVar(name="relogio_parede", vtype=GRB.CONTINUOUS, lb=0)

## Função objetivo: maximiza o lucro da venda dos dois tipos de relógio:

In [5]:
modelo.setObjective(relogio_pedestal["lucro"] * relogio_pedestal_var + relogio_parede["lucro"] * relogio_parede_var, GRB.MAXIMIZE)

## Restrições


* O número de horas gasto na montagem do mecanismo dos relógios não pode ultrapassar o número de horas máximo trabalhadas por João, que é o engenheiro:

In [6]:
modelo.addConstr(relogio_pedestal["tempo_gasto_montagem"] * relogio_pedestal_var + relogio_parede["tempo_gasto_montagem"] * relogio_parede_var <= horas_maximas_trabalhadas["Joao"], "horas_Joao")

<gurobi.Constr *Awaiting Model Update*>

* O número de horas gasto na produção dos relógios não pode ultrapassar o número de horas máximo trabalhadas por Ana, que é carpinteira:

In [7]:
modelo.addConstr(relogio_pedestal["tempo_gasto_producao"] * relogio_pedestal_var + relogio_parede["tempo_gasto_producao"] * relogio_parede_var <= horas_maximas_trabalhadas["Ana"], "horas_Ana")

<gurobi.Constr *Awaiting Model Update*>

* O número de horas gasto na realização de encomendas e remessas não pode ultrapassar o número máximo de horas trabalhadas por Lídia:

In [8]:
modelo.addConstr(relogio_pedestal["tempo_gasto_remessa"] * relogio_pedestal_var + relogio_parede["tempo_gasto_remessa"] * relogio_parede_var <= horas_maximas_trabalhadas["Lidia"], "horas_Lidia")

<gurobi.Constr *Awaiting Model Update*>

## Portanto, temos o seguinte modelo:

\begin{equation}
\max f(x,y) = 300x + 200y
\end{equation}

sujeito a:

\begin{alignat}{2}
6x + 4y \le 40 && \qquad \\
8x + 4y \le 40 && \qquad \\
3x + 3y \le 20 && \qquad \\
x \ge 0, y \ge 0 && \qquad
\end{alignat}

* Sendo x o número de relógios de pedestal e y o número de relógios de parede.

## Solução do modelo:

In [9]:
modelo.optimize()

Gurobi Optimizer version 12.0.0 build v12.0.0rc1 (linux64 - "Ubuntu 22.04.3 LTS")

CPU model: Intel(R) Xeon(R) CPU @ 2.20GHz, instruction set [SSE2|AVX|AVX2]
Thread count: 1 physical cores, 2 logical processors, using up to 2 threads

Optimize a model with 3 rows, 2 columns and 6 nonzeros
Model fingerprint: 0x6103ff60
Coefficient statistics:
  Matrix range     [3e+00, 8e+00]
  Objective range  [2e+02, 3e+02]
  Bounds range     [0e+00, 0e+00]
  RHS range        [2e+01, 4e+01]
Presolve time: 0.01s
Presolved: 3 rows, 2 columns, 6 nonzeros

Iteration    Objective       Primal Inf.    Dual Inf.      Time
       0    5.0000000e+32   4.250000e+30   5.000000e+02      0s
       2    1.6666667e+03   0.000000e+00   0.000000e+00      0s

Solved in 2 iterations and 0.01 seconds (0.00 work units)
Optimal objective  1.666666667e+03


## Impressão das soluções do modelo:

In [10]:
if(modelo.status == GRB.OPTIMAL):
  print(f"Relógio de pedestal: {relogio_pedestal_var.X}")
  print(f"Relógio de parede: {relogio_parede_var.X}")
  print(f"Valor ótimo: {modelo.objVal}")

Relógio de pedestal: 3.333333333333333
Relógio de parede: 3.3333333333333335
Valor ótimo: 1666.6666666666665


## 2$)$ Faça a análise de sensibilidade do modelo implementado, utilizando o gurobipy. Em sua resposta, além de apresentar os valores, apresente uma explicação sobre o que os resultados indicam:

In [11]:
print("\nIntervalos de Custos (Vetor de Custos):")
for variavel in modelo.getVars():
  print(f"Variável {variavel.VarName}:")
  print(f"  Lower bound do custo (SAObjLow): {variavel.SAObjLow}")
  print(f"  Upper bound do custo (SAObjUp): {variavel.SAObjUp}")

print("\nIntervalos de Recursos (Vetor de Recursos):")
for restricao in modelo.getConstrs():
  print(f"Restrição {restricao.ConstrName}:")
  print(f"  Lower bound do recurso (SARHSLow): {restricao.SARHSLow}")
  print(f"  Upper bound do recurso (SARHSUp): {restricao.SARHSUp}")
  print(f"  Variavel dual: {restricao.pi}")
  print(f"  Folga primal: {restricao.slack}")


Intervalos de Custos (Vetor de Custos):
Variável relogio_pedestal:
  Lower bound do custo (SAObjLow): 200.0
  Upper bound do custo (SAObjUp): 400.0
Variável relogio_parede:
  Lower bound do custo (SAObjLow): 150.0
  Upper bound do custo (SAObjUp): 300.0

Intervalos de Recursos (Vetor de Recursos):
Restrição horas_Joao:
  Lower bound do recurso (SARHSLow): 33.333333333333336
  Upper bound do recurso (SARHSUp): inf
  Variavel dual: 0.0
  Folga primal: 6.666666666666664
Restrição horas_Ana:
  Lower bound do recurso (SARHSLow): 26.666666666666668
  Upper bound do recurso (SARHSUp): 53.33333333333333
  Variavel dual: 25.0
  Folga primal: 0.0
Restrição horas_Lidia:
  Lower bound do recurso (SARHSLow): 15.0
  Upper bound do recurso (SARHSUp): 30.0
  Variavel dual: 33.333333333333336
  Folga primal: 0.0


##### Os valores de Lower Bound e Upper Bound indicam os valores tanto dos custos quanto das restições que podem ser admitidos para que a solução continue a mesma, isto é, que a partição básica e não básica contenham exatamente as mesmas variáveis.

##### A variável dual indica a variação da função objetivo caso aumente em uma unidade alguma restrição.

# 3$)$ Para aumentar o lucro total, os três sócios concordaram que um deles aumentaria ligeiramente o número máximo de horas disponíveis por semana. A escolha de qual deles se basearia naquele que aumentaria mais o lucro total. Use os resultados da análise de sensibilidade para fazer essa escolha:

##### A variável dual da Lídia (última restrição do modelo que está relacionada à remessa) possui o maior valor (i.e.: 33.333333333333336), então o seu aumento levaria ao meior aumento na função objetivo, pois para cada valor unitário de recurso adicionado, a função objetivo aumentaria em 33.333333333333336. Logo, deve-se aumentar o número de horas trabalhadas da Lídia.

# 4) Existe alguma variável dual igual a zero? O que significa ter uma variável dual igual a zero?

##### Sim, a variável dual do João (primeira restrição do modelo que está relacionada à montagem) possui valor zero, indicando que sua variação não varia o valor da função objetivo. Claro que caso a variação saia dos valores limitados pelo Upper bound e Lower bound teremos uma nova solução, modificando os valores das variáveis duais.

##### Nesse caso, o Upper bound é inf, logo pode-se aumentar infinitamente que não alterará o valor da função objetivo e a solução permanecerá a mesma.